File: src/main/java/featurecat/lizzie/util/KataGoAssetCatalog.java
package featurecat.lizzie.util;

import java.io.IOException;
import java.io.InputStream;
import java.nio.charset.StandardCharsets;
import java.util.Collections;
import java.util.LinkedHashMap;
import java.util.Map;
import org.json.JSONObject;

/** Trusted KataGo release assets shared by runtime setup and release packaging. */
public final class KataGoAssetCatalog {
  private static final String RESOURCE_NAME = "/katago-assets.json";
  private static final KataGoAssetCatalog INSTANCE = loadResource();

  private final String katagoVersion;
  private final String katagoReleaseTag;
  private final String katagoSourceCommit;
  private final String modelReleaseTag;
  private final String defaultModelId;
  private final Map<String, Model> models;
  private final Map<String, Asset> assets;

  private KataGoAssetCatalog(JSONObject root) {
    int schemaVersion = root.getInt("schemaVersion");
    if (schemaVersion != 1) {
      throw new IllegalStateException("Unsupported KataGo asset catalog schema: " + schemaVersion);
    }
    katagoVersion = required(root, "katagoVersion");
    katagoReleaseTag = required(root, "katagoReleaseTag");
    katagoSourceCommit = required(root, "katagoSourceCommit");
    modelReleaseTag = required(root, "modelReleaseTag");
    defaultModelId = required(root, "defaultModelId");
    models = Collections.unmodifiableMap(parseModels(root.getJSONObject("models")));
    assets = Collections.unmodifiableMap(parseAssets(root.getJSONObject("assets")));
    if (!models.containsKey(defaultModelId)) {
      throw new IllegalStateException("Unknown default KataGo model: " + defaultModelId);
    }
    Asset windowsNvidia = assets.get("windows-nvidia");
    if (windowsNvidia == null || windowsNvidia.executableSha256().isEmpty()) {
      throw new IllegalStateException(
          "KataGo asset catalog requires windows-nvidia executableSha256");
    }
    if (!assets.containsKey("windows-tensorrt")) {
      throw new IllegalStateException("KataGo asset catalog requires windows-tensorrt");
    }
  }

  public static KataGoAssetCatalog get() {
    return INSTANCE;
  }

  public String katagoVersion() {
    return katagoVersion;
  }

  public String katagoReleaseTag() {
    return katagoReleaseTag;
  }

  public String katagoSourceCommit() {
    return katagoSourceCommit;
  }

  public String modelReleaseTag() {
    return modelReleaseTag;
  }

  public Model defaultModel() {
    return model(defaultModelId);
  }

  public Model model(String id) {
    Model model = models.get(id);
    if (model == null) {
      throw new IllegalArgumentException("Unknown KataGo model id: " + id);
    }
    return model;
  }

  public Asset asset(String id) {
    Asset asset = assets.get(id);
    if (asset == null) {
      throw new IllegalArgumentException("Unknown KataGo asset id: " + id);
    }
    return asset;
  }

  public Map<String, Model> models() {
    return models;
  }

  public Map<String, Asset> assets() {
    return assets;
  }

  public String modelDownloadUrl(Model model) {
    return model.downloadUrl().isEmpty()
        ? releaseUrl(modelReleaseTag, model.fileName())
        : model.downloadUrl();
  }

  public String assetDownloadUrl(Asset asset) {
    return releaseUrl(katagoReleaseTag, asset.assetName());
  }

  private static String releaseUrl(String tag, String fileName) {
    return "https://github.com/lightvector/KataGo/releases/download/" + tag + "/" + fileName;
  }

  private static Map<String, Model> parseModels(JSONObject values) {
    Map<String, Model> parsed = new LinkedHashMap<>();
    for (String id : values.keySet()) {
      JSONObject value = values.getJSONObject(id);
      parsed.put(
          id,
          new Model(
              id,
              required(value, "fileName"),
              required(value, "displayFamily"),
              required(value, "tier"),
              required(value, "architecture"),
              required(value, "minimumKataGoVersion"),
              value.getLong("sizeBytes"),
              requiredSha256(value, "sha256"),
              value.optBoolean("bundled", false),
              modelDownloadOverride(value),
              value.optString("publishedAt", "2026-07-29")));
    }
    return parsed;
  }

  private static String modelDownloadOverride(JSONObject value) {
    String url = value.optString("downloadUrl", "").trim();
    if (!url.isEmpty()
        && !url.equals(
            "https://media.katagotraining.org/uploaded/networks/models/kata1/"
                + required(value, "fileName"))) {
      throw new IllegalStateException("Unsupported official model download URL");
    }
    return url;
  }

  private static Map<String, Asset> parseAssets(JSONObject values) {
    Map<String, Asset> parsed = new LinkedHashMap<>();
    for (String id : values.keySet()) {
      JSONObject value = values.getJSONObject(id);
      parsed.put(
          id,
          new Asset(
              id,
              required(value, "platform"),
              required(value, "backend"),
              required(value, "assetName"),
              value.getLong("sizeBytes"),
              requiredSha256(value, "sha256"),
              optionalSha256(value, "executableSha256"),
              value.optString("runtimeProfile", ""),
              value.optString("gpuFamily", ""),
              required(value, "releaseTier")));
    }
    return parsed;
  }

  private static String required(JSONObject value, String key) {
    String result = value.optString(key, "").trim();
    if (result.isEmpty()) {
      throw new IllegalStateException("Missing KataGo asset catalog field: " + key);
    }
    return result;
  }

  private static String requiredSha256(JSONObject value, String key) {
    String sha256 = required(value, key).toLowerCase();
    validateSha256(key, sha256);
    return sha256;
  }

  private static String optionalSha256(JSONObject value, String key) {
    String sha256 = value.optString(key, "").trim().toLowerCase();
    if (!sha256.isEmpty()) {
      validateSha256(key, sha256);
    }
    return sha256;
  }

  private static void validateSha256(String key, String sha256) {
    if (!sha256.matches("[0-9a-f]{64}")) {
      throw new IllegalStateException("Invalid KataGo asset catalog SHA-256 for " + key);
    }
  }

  private static KataGoAssetCatalog loadResource() {
    try (InputStream input = KataGoAssetCatalog.class.getResourceAsStream(RESOURCE_NAME)) {
      if (input == null) {
        throw new IllegalStateException("Missing bundled KataGo asset catalog: " + RESOURCE_NAME);
      }
      String json = new String(input.readAllBytes(), StandardCharsets.UTF_8);
      return new KataGoAssetCatalog(new JSONObject(json));
    } catch (IOException e) {
      throw new IllegalStateException("Failed to load bundled KataGo asset catalog", e);
    }
  }

  public record Model(
      String id,
      String fileName,
      String displayFamily,
      String tier,
      String architecture,
      String minimumKataGoVersion,
      long sizeBytes,
      String sha256,
      boolean bundled,
      String downloadUrl,
      String publishedAt) {
    public String modelName() {
      return fileName.endsWith(".bin.gz")
          ? fileName.substring(0, fileName.length() - ".bin.gz".length())
          : fileName;
    }
  }

  public record Asset(
      String id,
      String platform,
      String backend,
      String assetName,
      long sizeBytes,
      String sha256,
      String executableSha256,
      String runtimeProfile,
      String gpuFamily,
      String releaseTier) {}
}

File: scripts/katago_asset_catalog.py
#!/usr/bin/env python3
"""Read and validate the trusted KataGo release asset catalog."""

from __future__ import annotations

import argparse
import json
import re
from pathlib import Path
from typing import Any

ROOT = Path(__file__).resolve().parents[1]
DEFAULT_CATALOG = ROOT / "src" / "main" / "resources" / "katago-assets.json"
SHA256_RE = re.compile(r"^[0-9a-f]{64}$")


def load_catalog(path: Path) -> dict[str, Any]:
    with path.open("r", encoding="utf-8") as handle:
        catalog = json.load(handle)
    validate_catalog(catalog)
    return catalog


def validate_catalog(catalog: dict[str, Any]) -> None:
    if catalog.get("schemaVersion") != 1:
        raise ValueError("KataGo asset catalog schemaVersion must be 1")
    version = require_text(catalog, "katagoVersion")
    release_tag = require_text(catalog, "katagoReleaseTag")
    if release_tag != f"v{version}":
        raise ValueError("katagoReleaseTag must match katagoVersion")
    models = catalog.get("models")
    assets = catalog.get("assets")
    if not isinstance(models, dict) or not models:
        raise ValueError("models must be a non-empty object")
    if not isinstance(assets, dict) or not assets:
        raise ValueError("assets must be a non-empty object")
    default_id = require_text(catalog, "defaultModelId")
    if default_id not in models:
        raise ValueError(f"defaultModelId is unknown: {default_id}")
    bundled = [model_id for model_id, model in models.items() if model.get("bundled") is True]
    if bundled != [default_id]:
        raise ValueError("exactly the default model must be marked bundled")
    for model_id, model in models.items():
        validate_entry(model, f"model {model_id}")
        require_text(model, "fileName")
        require_text(model, "minimumKataGoVersion")
        override = model.get("downloadUrl", "")
        if override and override != (
            "https://media.katagotraining.org/uploaded/networks/models/kata1/" + model["fileName"]
        ):
            raise ValueError(f"model {model_id} has unsupported official downloadUrl")
    for asset_id, asset in assets.items():
        validate_entry(asset, f"asset {asset_id}")
        name = require_text(asset, "assetName")
        if f"-{release_tag}-" not in name:
            raise ValueError(f"asset {asset_id} does not use {release_tag}: {name}")
        executable_sha = asset.get("executableSha256", "")
        if executable_sha and not SHA256_RE.fullmatch(executable_sha):
            raise ValueError(f"asset {asset_id} has invalid executableSha256")
    windows_nvidia = assets.get("windows-nvidia", {})
    if not SHA256_RE.fullmatch(str(windows_nvidia.get("executableSha256", ""))):
        raise ValueError("asset windows-nvidia requires executableSha256")
    if "windows-tensorrt" not in assets:
        raise ValueError("asset windows-tensorrt is required")


def validate_entry(entry: Any, label: str) -> None:
    if not isinstance(entry, dict):
        raise ValueError(f"{label} must be an object")
    if not isinstance(entry.get("sizeBytes"), int) or entry["sizeBytes"] <= 0:
        raise ValueError(f"{label} has invalid sizeBytes")
    sha256 = require_text(entry, "sha256")
    if not SHA256_RE.fullmatch(sha256):
        raise ValueError(f"{label} has invalid sha256")


def require_text(value: dict[str, Any], key: str) -> str:
    result = value.get(key)
    if not isinstance(result, str) or not result.strip():
        raise ValueError(f"missing non-empty string field: {key}")
    return result.strip()


def resolve(catalog: dict[str, Any], dotted_path: str) -> Any:
    current: Any = catalog
    for part in dotted_path.split("."):
        if not isinstance(current, dict) or part not in current:
            raise KeyError(f"unknown catalog path: {dotted_path}")
        current = current[part]
    return current


def model_download_url(catalog: dict[str, Any], model_id: str) -> str:
    model = catalog["models"][model_id]
    return model.get("downloadUrl") or (
        f"https://github.com/lightvector/KataGo/releases/download/{catalog['modelReleaseTag']}/"
        + model["fileName"]
    )


def main() -> int:
    parser = argparse.ArgumentParser()
    parser.add_argument("--catalog", type=Path, default=DEFAULT_CATALOG)
    subparsers = parser.add_subparsers(dest="command", required=True)
    subparsers.add_parser("validate")
    get_parser = subparsers.add_parser("get")
    get_parser.add_argument("path")
    model_url_parser = subparsers.add_parser("model-url")
    model_url_parser.add_argument("model_id")
    args = parser.parse_args()

    catalog = load_catalog(args.catalog)
    if args.command == "model-url":
        print(model_download_url(catalog, args.model_id))
    if args.command == "get":
        value = resolve(catalog, args.path)
        if isinstance(value, (dict, list)):
            print(json.dumps(value, ensure_ascii=False, separators=(",", ":")))
        elif isinstance(value, bool):
            print("true" if value else "false")
        else:
            print(value)
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
